### PR TITLE
backend/local: check and return iter.Error when pebble is not valid

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
Check pebble iter.Error when iter is not valid so if iter returns an error (such as caused by exceeding max open file limit), lightning won't omit the actual error

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

Related changes

### CODE REVIEW

File: lightning/backend/local.go [hunk 1]
```diff
@@ -705,16 +709,24 @@ func (local *local) readAndSplitIntoRange(engineFile *LocalFile) ([]Range, error
 	iter := engineFile.db.NewIter(nil)
 	defer iter.Close()
 
+	iterError := func(e string) error {
+		err := iter.Error()
+		if err != nil {
+			return errors.Annotatef(err, e)
```
Comment (kennytm) [inline lightning/backend/local.go:715]:
```suggestion
			return errors.Annotate(err, e)
```

File: lightning/backend/local.go [hunk 2]
```diff
@@ -500,8 +500,12 @@ func (local *local) WriteToTiKV(
 	opt := &pebble.IterOptions{LowerBound: regionRange.start, UpperBound: regionRange.end}
 	iter := engineFile.db.NewIter(opt)
 	defer iter.Close()
+	iter.First()
+	if iter.Error() != nil {
+		return nil, nil, errors.Annotatef(iter.Error(), "failed to read the first key")
```
Comment (kennytm) [inline lightning/backend/local.go:505]:
```suggestion
		return nil, nil, errors.Annotate(iter.Error(), "failed to read the first key")
```

File: lightning/backend/local.go [hunk 3]
```diff
@@ -851,6 +863,9 @@ func (local *local) writeAndIngestByRange(
 	defer iter.Close()
 	// Needs seek to first because NewIter returns an iterator that is unpositioned
 	hasKey := iter.First()
+	if iter.Error() != nil {
+		return errors.Annotatef(iter.Error(), "failed to read the first key")
```
Comment (kennytm) [inline lightning/backend/local.go:867]:
```suggestion
		return errors.Annotate(iter.Error(), "failed to read the first key")
```

**Review by kennytm — COMMENTED** [review]
Rest LGTM

**Comment by kennytm** [comment]
LGTM

**Comment by kennytm** [comment]
/test

```
[2020-11-30T10:22:17.595Z] 
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed connect to fileserver.pingcap.net:80; Connection refused
```

File: lightning/backend/local.go [hunk 4]
```diff
@@ -500,8 +500,12 @@ func (local *local) WriteToTiKV(
 	opt := &pebble.IterOptions{LowerBound: regionRange.start, UpperBound: regionRange.end}
 	iter := engineFile.db.NewIter(opt)
 	defer iter.Close()
+	iter.First()
+	if iter.Error() != nil {
+		return nil, nil, errors.Annotate(iter.Error(), "failed to read the first key")
+	}
 
-	if !iter.First() {
+	if !iter.Valid() {
```
Comment (overvenus) [inline lightning/backend/local.go:508]: What about `iter.Last()` at line 516?

File: lightning/backend/local.go [hunk 5]
```diff
@@ -851,6 +863,9 @@ func (local *local) writeAndIngestByRange(
 	defer iter.Close()
 	// Needs seek to first because NewIter returns an iterator that is unpositioned
 	hasKey := iter.First()
+	if iter.Error() != nil {
+		return errors.Annotate(iter.Error(), "failed to read the first key")
+	}
```
Comment (overvenus) [inline lightning/backend/local.go:868]: What about iter.Last() at line 877?

Comment (overvenus) [reply lightning/backend/local.go:871]: What about this comment? @glorv

Comment (glorv) [reply lightning/backend/local.go:871]: fixed

**Comment by overvenus** [comment]
@glorv Please make sure this PR adds all necessary error checks for pebble usage.

**Comment by glorv** [comment]
> @glorv Please make sure this PR adds all necessary error checks for pebble usage.

I have checked all the pebble iterators, there should be no remain.

File: lightning/backend/local.go (hunk 5 above)
Comment (lance6716) [reply lightning/backend/local.go:871]: do we need to check `Valid()` after `iter.Last()`, as we did for `iter.First()`?

Comment (glorv) [reply lightning/backend/local.go:871]: yes,  see #880-881 @lance6716

Comment (lance6716) [reply lightning/backend/local.go:871]: Do you mean `iter.Error()` here could cover `iter.Valid()`? For `iter.First()` why we need them both (maybe iterator will keep valid after normal operation on iterator 🤔 )

Comment (glorv) [reply lightning/backend/local.go:871]: we only need to check iter.Error() != nil which  indicates that something goes wrong. iter.Error() != nil always means iter.Valid() is false. if iter.Error() == nil and iter.Valid() is false, this means the iterator iters to the end, this is ok to our logic.

**Comment by lance6716** [comment]
LGTM

**Review by lance6716 — APPROVED** [review]

**Pull request merged**